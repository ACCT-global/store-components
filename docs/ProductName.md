ðŸ“¢ Don't fork this project. Use, [contribute](https://github.com/vtex-apps/awesome-io#contributing), or open issues through [Store Discussion](https://github.com/vtex-apps/store-discussion).

# Product Name

`ProductName` is a VTEX Component that shows the name of the product with other informations such as SKU or brand.
This Component can be imported and used by any VTEX App.

![image](https://user-images.githubusercontent.com/284515/70231165-8f6b4200-1738-11ea-9f06-3583c08fc693.png)

## Configuration

1. Import the vtex.store-component's app to your theme's dependencies in the manifest.json, for example:

```json
  dependencies: {
    "vtex.store-components": "3.x"
  }
```

2. Add the `product-name` block to any block bellow `store.product`. For example:

```json
  "store.product": {
    "children": [
      "flex-layout.row#product",
    ]
  },
  "flex-layout.row#product": {
    "children": [
      "product-name"
    ]
  }
```

| Prop name | Type | Description |
| --- | --- | --- |
| `showSku` | `Boolean` | Show product SKU |
| `showProductReference` | `Boolean` | Show product reference |
| `showBrandName` | `Boolean` | Show brand name |

## Customization

| CSS Handles |
| --- |
| `productNameContainer` |
| `productBrand` |
| `productSku` |
| `productReference` |
| `productNameLoader` |
| `productNameBrandLoader` |
| `productNameSkuLoader` |