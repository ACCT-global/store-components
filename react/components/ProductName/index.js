import PropTypes from 'prop-types'
import React, { Component } from 'react'
import ContentLoader from 'react-content-loader'
import classNames from 'classnames'

/**
 * Name component. Show name and relevant SKU information of the Product Summary
 */
class ProductName extends Component {
  static propTypes = {
    /** Name of the product */
    name: PropTypes.string,
    /** Selected SKU name */
    skuName: PropTypes.string,
    /** Show sku */
    showSku: PropTypes.bool,
    /** Product reference */
    productReference: PropTypes.string,
    /** Show product reference */
    showProductReference: PropTypes.bool,
    /** Brand name */
    brandName: PropTypes.string,
    /** Show brand name */
    showBrandName: PropTypes.bool,
    /** Display large font */
    large: PropTypes.bool,
    /** Extra classes for loader of the component */
    loaderClasses: PropTypes.string,
    /** Extra classes for root of the component */
    rootClasses: PropTypes.string,
    /** Component and content loader styles */
    styles: PropTypes.object,
    /** Render function for name field, this function should expect a name string and large prop */
    renderName: PropTypes.func,
    /** Render function for sku field, this function should expect a skuName string and large prop */
    renderSku: PropTypes.func,
    /** Render function for product reference field, this funcitno should expect a productReference string and large prop */
    renderProductReference: PropTypes.func
  }

  static defaultProps = {
    large: false,
    showBrandName: false,
    showProductReference: false,
    showSku: false,
    loaderClasses: '',
    rootClasses: ''
  }

  static Loader = (loaderProps = {}) => (
    <div className={loaderProps.loaderClasses}>
      <ContentLoader
        style={{
          width: '100%',
          height: '100%',
        }}
        height="100%"
        width="100%"
        {...loaderProps}
      >
        <rect
          height="1.125em"
          width="75%"
          x="15%"
          {...loaderProps['vtex-product-name__brand--loader']}
        />
        <rect
          height="1.125em"
          width="50%"
          x="25%"
          y="1.75em"
          {...loaderProps['vtex-product-name__sku--loader']}
        />
      </ContentLoader>
    </div>
  )

  render() {
    const {
      name,
      large,
      skuName,
      showSku,
      renderName,
      rootClasses,
      loaderClasses,
      productReference,
      showProductReference,
    } = this.props

    if (!name) {
      return (
        <ProductName.Loader
          loaderClasses={classNames('vtex-product-name', 'vtex-product-name-loader', loaderClasses)}
          {...this.props.styles}
        />
      )
    }

    return (
      <div className={classNames('vtex-product-name', rootClasses)}>
        {renderName && renderName({ large, name })}
        {showSku && renderSku && renderSku({ skuName, large })}
        {showProductReference && renderProductReference
          && renderProductReference({ productReference, large })}
      </div>
    )
  }
}

ProductName.schema = {
  title: 'editor.productName.title',
  description: 'editor.productName.description',
  type: 'object',
  properties: {
    showBrandName: {
      type: 'boolean',
      title: 'editor.productName.showBrandName.title',
      default: false,
      isLayout: true,
    },
    showSku: {
      type: 'boolean',
      title: 'editor.productName.showSku.title',
      default: false,
      isLayout: true,
    },
    showProductReference: {
      type: 'boolean',
      title: 'editor.productName.showProductReference.title',
      default: false,
      isLayout: true,
    },
  },
}

export default ProductName
