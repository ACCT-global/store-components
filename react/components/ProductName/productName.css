/**
    ProductName CSS Module
*/
.container {}
.loader {}
.brand {}
.sku {}
.productReference {}
