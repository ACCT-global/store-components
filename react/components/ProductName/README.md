# Product Name
Product Name is a canonical component that any VTEX app can import.

To import it into your code: 
```js
import ProductName from 'vtex.store-components/ProductName'
```

## Usage
You can use it in your code like a React component with the jsx tag: `<ProductName />`. 
```jsx
// ...

<ProductName
  name={product.productName}
  skuName={selectedItem.name}
  brandName={product.brand}
  renderName={({ name, large }) => {
      let classes = 'vtex-product-name__brand c-on-base'
      if (large) {
        classes = classNames(classes, 'some-css-class-for-large-typograhy')
      } else {
        classes = classNames(classes, 'some-css-class-for-small-typography')
      }
      return (
        <span className={classes}>{name}</span>
      )
  }}
/>

// The others render props functions should be very similar with this one
```

| Prop name               | Type       | Description                                                                                                       |
| ----------------------- | ---------- | ----------------------------------------------------------------------------------------------------------------- |
| `brandName`             | `String`   | Brand name                                                                                                        |
| `large`                 | `Boolean`  | Displays large font                                                                                               |
| `name`                  | `String!`  | Name of the product                                                                                               |
| `productReference`      | `String`   | Product reference                                                                                                 |
| `renderName`            | `func`     | Render function for name field, this function should expect a name string and large prop                          |
| `renderSku`             | `func`     | Render function for sku field, this function should expect a skuName string and large prop                        |
| `renderProductReference`| `func`     | Render function for product reference field, this function should expect a productReference string and large prop |
| `showBrandName`         | `Boolean`  | Show brand name                                                                                                   |
| `showProductReference`  | `Boolean`  | Show product reference                                                                                            |
| `showSku`               | `Boolean`  | Show product SKU                                                                                                  |
| `skuName`               | `String`   | Selected SKU name                                                                                                 |

See an example at [Product Details](https://github.com/vtex-apps/product-summary/blob/master/react/index.js) app
