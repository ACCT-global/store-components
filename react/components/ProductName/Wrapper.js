import React, { useContext } from 'react'
import { path, isEmpty } from 'ramda'
import { ProductContext } from 'vtex.product-context'

import ProductName from './index'

const styles = {
  'vtex-product-name__brand--loader': {
    x: 0,
    width: '100%',
    height: '1.631em',
  },
  'vtex-product-name__sku--loader': {
    x: 0,
    y: '2.569em',
    width: '10.311em',
    height: '1.045em',
  },
}

const ProductNameWrapper = props => {
  const { showBrandName, showSku, showProductReference } = props

  const valuesFromContext = useContext(ProductContext)

  const productNameProps = () => {
    if (!valuesFromContext || isEmpty(valuesFromContext)) {
      return props
    }

    const { product, selectedItem } = valuesFromContext
    return {
      tag: 'h1',
      name: path(['productName'], product),
      skuName: path(['name'], selectedItem),
      productReference: path(['productReference'], product),
      brandName: path(['brand'], product),
      styles: styles,
      className: 't-heading-4',
      showBrandName,
      showSku,
      showProductReference,
    }
  }

  return <ProductName {...productNameProps()} />
}

ProductNameWrapper.schema = {
  title: 'admin/editor.productName.title',
  description: 'admin/editor.productName.description',
  type: 'object',
  properties: {
    showBrandName: {
      type: 'boolean',
      title: 'admin/editor.productName.showBrandName.title',
      default: false,
      isLayout: true,
    },
    showSku: {
      type: 'boolean',
      title: 'admin/editor.productName.showSku.title',
      default: false,
      isLayout: true,
    },
    showProductReference: {
      type: 'boolean',
      title: 'admin/editor.productName.showProductReference.title',
      default: false,
      isLayout: true,
    },
  },
}

export default ProductNameWrapper
