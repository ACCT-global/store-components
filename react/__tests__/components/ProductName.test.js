import React from 'react'
import ProductName from './../../ProductName'
import { render } from 'enzyme'
import PropTypes from 'prop-types'

describe('<ProductName />', () => {
  const defaultProps = {
    loaderClass: '',
    name: 'ProductTest',
  }

  const context = { culture: { currency: 'USD' } }

  const renderComponent = customProps => {
    const props = {
      ...defaultProps,
      ...customProps,
    }
    return render(<ProductName {...props} />, {
      context,
      childContextTypes: {
        culture: PropTypes.object,
      },
    })
  }

  it('should be mount', () => {
    expect(renderComponent()).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(renderComponent()).toMatchSnapshot()
  })

  it('should match snapshot with SkuName', () => {
    expect(
      renderComponent({ skuName: 'ProductSkuName', showSku: true })
    ).toMatchSnapshot()
  })

  it('should match snapshot Loader', () => {
    expect(renderComponent({ name: undefined })).toMatchSnapshot()
  })
})
