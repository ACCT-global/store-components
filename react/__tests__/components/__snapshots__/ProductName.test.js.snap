// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductName /> should match snapshot 1`] = `
<div
  class="productNameContainer mv0"
>
  <span
    class="productBrand"
  >
    ProductTest 
  </span>
</div>
`;

exports[`<ProductName /> should match snapshot Loader 1`] = `
<div
  class="productNameContainer productNameLoader"
>
  <div
    class="content-loader-mock"
  >
    <rect
      height="1.125em"
      width="75%"
      x="15%"
    />
    <rect
      height="1.125em"
      width="50%"
      x="25%"
      y="1.75em"
    />
  </div>
</div>
`;

exports[`<ProductName /> should match snapshot with SkuName 1`] = `
<div
  class="productNameContainer mv0"
>
  <span
    class="productBrand"
  >
    ProductTest 
  </span>
  <span
    class="productBrand"
  >
    ProductSkuName
  </span>
</div>
`;
